### ISSUE TITLE
"Install Dev Extension" silently fails because `rustup` is not installed

### ISSUE THREAD

**Issue by ydalton** [zed-industries/zed#21237]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

If you try to install a dev extension, if `rustup` is not installed, then the installation will silently fail (e.g. no error message). The problem is that error is _only_ visible in the logs, and doesn't show a message saying what went wrong.

Steps to reproduce:
- Make sure that `rustup` is not installed.
- Go to the extensions page and click on "Install Dev Extension".
- Select the path where a dev extension repo is.
- The extension will not install.
- Check the logs to verify that it could not install because of a missing `rustup` (see below)

### Environment

Zed: v0.162.5 (Zed)
OS: Linux Wayland gentoo 2.17
Memory: 15.3 GiB
Architecture: x86_64
GPU: Intel(R) Graphics (RPL-P) || Intel open-source Mesa driver || Mesa 24.2.6

### If applicable, attach your Zed.log file to this issue.

```

```
</details>
```
[2024-11-27T13:53:57+01:00 ERROR extensions_ui] failed to compile Rust extension

Caused by:
    0: failed to run `rustup target add`
    1: No such file or directory (os error 2)
```

**Comment by notpeter** [comment]
I'm with you. It'd be nice to pop-up an error message perhaps with a link to:
- https://zed.dev/docs/extensions/developing-extensions

I've adjusted the docs on that page so the rustup callout is at the top:
- https://github.com/zed-industries/zed/pull/21239

**Comment by WeetHet** [comment]
You don't _really_ need rustup though. What you need is a matching target which is wasip1 [for now](#20811). You can get around with just using whatever other method of installation you're using for as long as it installs that target, as zed doesn't indiscriminately install rustup. For example, rust-overlay for nix can install different targets as well, so it also works. Therefore, the callout shouldn't be that rustup is needed per se, it should be more of: wasip{1/2} target should be available OR rustup should be installed

**Comment by ydalton** [comment]
> I'm with you. It'd be nice to pop-up an error message perhaps with a link to:
> 
> * https://zed.dev/docs/extensions/developing-extensions

Just to clarify, I'm not a Zed extension developer, I just wanted to install an extension which was on GitHub and not in the official Zed extension listing.

**Comment by uncenter** [comment]
Yeah this is super frustrating. I don't have rustup installed since I use Rust toolchains through Nix, and so installing extensions with LSPs locally fail silently as mentioned because of this. Definitely a two part issue here, in that a) there should not be silent failures and b) rustup shouldn't be required as mentioned by @WeetHet.

**Comment by notpeter** [comment]
> Therefore, the callout shouldn't be that rustup is needed per se, it should be more of: wasip{1/2} target should be available OR rustup should be installed.

@WeetHet Happy to review PRs to improve docs precision here.

The other reason that I think this hasn't improved is that it's actually kind of a PITA for Zed developers to trigger this behavior:
1. To try and reproduce/test I need to uninstall rustup/toolchains on my mac
2. To make a change that improves that error messaging, I need to install rustup/toolchains and compile Zed from scratch

So the iteration cycle would be a minimum of 5-10minutes for each go round as I break/fix my dev environment.  It's totally worth fixing, but I also understand why no one yet is rushing to do the work.

**Comment by uncenter** [comment]
> To try and reproduce/test I need to uninstall rustup/toolchains on my mac

I don't think reproduction is needed. You could just insert a manually throw an error at that spot in the code regardless of whether rustup is installed, and then uninstalling/reinstalling rustup wouldn't be an issue I think?

**Comment by Swiftaff** [comment]
Hi, I 've been using Zed for a while and was looking for issues to start contributing to, and I'd like to help out on this one if possible.

I have a draft PR working https://github.com/zed-industries/zed/compare/main...Swiftaff:zed:21237_pr which addresses the pop-up error message part for now, but I wasn't sure on how or if a popup would work so I suggest a pseudo extension card at the top of the list, in order to give the user immediate feedback that the install has started as it takes quite a few seconds to compile some extensions for me without any indication it was still in progress.

@uncenter suggestion helped me easily trigger the rustc error, or you can invalidate the extension.toml too, and these errors are then shown in the closeable card with a link to the suggested webpage. Let me know if this is overkill, or if useful how to proceed if any changes/tests needed, my RUST is ok, but e.g. not sure how to address @WeetHet point about rustup not being required.

![dev_extension_cards](https://github.com/user-attachments/assets/0d7ea668-f8e0-4239-b4b0-ba3de84a7026)

**Comment by uncenter** [comment]
I would prefer a toast notification in the bottom right to a card in the extension page like you have it now I think. If I install the dev extension from the command palette, I won't be on the extensions page at all so I wouldn't see it.

**Comment by Swiftaff** [comment]
Ah thanks yeh that makes sense. OK, I removed the custom cards and got toasts working, including from the command palette, but currently only when the extension_ui is visible as they are triggered from there. I'm now looking into how/which view to trigger the toasts from more globally.
